<a href="https://colab.research.google.com/github/bethmorais/aed-edd/blob/master/aed_edd_divorce_task2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Equipe: Elisabeth Wanderley e Lucas

### **Importação da Base e Primeiros Ajustes** 

Os dados foram coletados a partir de https://archive.ics.uci.edu/ml/datasets/Divorce+Predictors+data+set, tendo sido exportados para uma planilha csv, disponibilizada no git do projeto.

In [1]:
import numpy as np
import os
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, KFold
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report, confusion_matrix, precision_score

dataset = pd.read_csv('https://raw.githubusercontent.com/bethmorais/aed-edd/master/divorce.csv')
dataset.head()

Atr1  Atr2  Atr3  Atr4  Atr5  Atr6  ...  Atr50  Atr51  Atr52  Atr53  Atr54  Class
0     2     2     4     1     0     0  ...      3      2      3      2      1      1
1     4     4     4     4     4     0  ...      4      4      4      2      2      1
2     2     2     2     2     1     3  ...      1      1      2      2      2      1
3     3     2     3     2     3     3  ...      3      3      2      2      2      1
4     2     2     1     1     1     1  ...      2      2      2      1      0      1

[5 rows x 55 columns]

Verificação dos tipos de dados inferidos pelo pandas

In [2]:
dataset.dtypes

Atr1     int64
Atr2     int64
Atr3     int64
Atr4     int64
Atr5     int64
Atr6     int64
Atr7     int64
Atr8     int64
Atr9     int64
Atr10    int64
Atr11    int64
Atr12    int64
Atr13    int64
Atr14    int64
Atr15    int64
Atr16    int64
Atr17    int64
Atr18    int64
Atr19    int64
Atr20    int64
Atr21    int64
Atr22    int64
Atr23    int64
Atr24    int64
Atr25    int64
Atr26    int64
Atr27    int64
Atr28    int64
Atr29    int64
Atr30    int64
Atr31    int64
Atr32    int64
Atr33    int64
Atr34    int64
Atr35    int64
Atr36    int64
Atr37    int64
Atr38    int64
Atr39    int64
Atr40    int64
Atr41    int64
Atr42    int64
Atr43    int64
Atr44    int64
Atr45    int64
Atr46    int64
Atr47    int64
Atr48    int64
Atr49    int64
Atr50    int64
Atr51    int64
Atr52    int64
Atr53    int64
Atr54    int64
Class    int64
dtype: object

Verificando a quantidade de instâncias (linhas) e atributos (colunas) antes de limpar os dados

In [3]:
dataset.shape

(170, 55)

*Verificando* se há respostas com atributos nulos

In [4]:
print(dataset.isnull().sum())

Atr1     0
Atr2     0
Atr3     0
Atr4     0
Atr5     0
Atr6     0
Atr7     0
Atr8     0
Atr9     0
Atr10    0
Atr11    0
Atr12    0
Atr13    0
Atr14    0
Atr15    0
Atr16    0
Atr17    0
Atr18    0
Atr19    0
Atr20    0
Atr21    0
Atr22    0
Atr23    0
Atr24    0
Atr25    0
Atr26    0
Atr27    0
Atr28    0
Atr29    0
Atr30    0
Atr31    0
Atr32    0
Atr33    0
Atr34    0
Atr35    0
Atr36    0
Atr37    0
Atr38    0
Atr39    0
Atr40    0
Atr41    0
Atr42    0
Atr43    0
Atr44    0
Atr45    0
Atr46    0
Atr47    0
Atr48    0
Atr49    0
Atr50    0
Atr51    0
Atr52    0
Atr53    0
Atr54    0
Class    0
dtype: int64


In [0]:
df = dataset.describe()

In [6]:
dfc1 = dataset[(dataset['Class'] == 1)]
dfc1.describe()

Atr1       Atr2       Atr3  ...      Atr53      Atr54  Class
count  84.000000  84.000000  84.000000  ...  84.000000  84.000000   84.0
mean    3.190476   2.869048   2.916667  ...   3.321429   3.369048    1.0
std     0.752125   0.915460   0.747821  ...   0.920460   1.095223    0.0
min     0.000000   0.000000   1.000000  ...   1.000000   0.000000    1.0
25%     3.000000   2.000000   3.000000  ...   3.000000   3.000000    1.0
50%     3.000000   3.000000   3.000000  ...   4.000000   4.000000    1.0
75%     4.000000   4.000000   3.000000  ...   4.000000   4.000000    1.0
max     4.000000   4.000000   4.000000  ...   4.000000   4.000000    1.0

[8 rows x 55 columns]

In [7]:
dfc1.shape

(84, 55)

# **1o Exercício**

### 11/Janeiro/2020
#### Identificar que perguntas são mais significativas na base de divórcio 

In [8]:
dfc0 = dataset[(dataset['Class'] == 0)]
dfc0.describe()

Atr1       Atr2       Atr3  ...      Atr53      Atr54  Class
count  86.000000  86.000000  86.000000  ...  86.000000  86.000000   86.0
mean    0.395349   0.465116   0.639535  ...   1.186047   0.686047    0.0
std     0.898090   0.762313   0.919243  ...   1.183332   0.871262    0.0
min     0.000000   0.000000   0.000000  ...   0.000000   0.000000    0.0
25%     0.000000   0.000000   0.000000  ...   0.000000   0.000000    0.0
50%     0.000000   0.000000   0.000000  ...   1.000000   0.000000    0.0
75%     0.000000   1.000000   1.000000  ...   2.000000   1.000000    0.0
max     4.000000   3.000000   4.000000  ...   4.000000   4.000000    0.0

[8 rows x 55 columns]

In [9]:
dfc0.shape

(86, 55)

In [10]:
dfc0_mean = dfc0.mean()
dfc0_mean

Atr1     0.395349
Atr2     0.465116
Atr3     0.639535
Atr4     0.267442
Atr5     0.104651
Atr6     0.372093
Atr7     0.011628
Atr8     0.127907
Atr9     0.058140
Atr10    0.406977
Atr11    0.197674
Atr12    0.395349
Atr13    0.604651
Atr14    0.290698
Atr15    0.232558
Atr16    0.162791
Atr17    0.174419
Atr18    0.093023
Atr19    0.139535
Atr20    0.069767
Atr21    0.151163
Atr22    0.069767
Atr23    0.081395
Atr24    0.267442
Atr25    0.337209
Atr26    0.197674
Atr27    0.151163
Atr28    0.081395
Atr29    0.093023
Atr30    0.197674
Atr31    0.837209
Atr32    0.732558
Atr33    0.290698
Atr34    0.558140
Atr35    0.104651
Atr36    0.034884
Atr37    0.627907
Atr38    0.348837
Atr39    0.569767
Atr40    0.209302
Atr41    0.476744
Atr42    1.011628
Atr43    1.953488
Atr44    0.534884
Atr45    1.651163
Atr46    1.953488
Atr47    1.244186
Atr48    2.046512
Atr49    1.279070
Atr50    1.383721
Atr51    1.616279
Atr52    1.569767
Atr53    1.186047
Atr54    0.686047
Class    0.000000
dtype: flo

In [11]:
dfc1_mean = dfc1.mean()
dfc1_mean

Atr1     3.190476
Atr2     2.869048
Atr3     2.916667
Atr4     2.726190
Atr5     3.011905
Atr6     1.130952
Atr7     0.988095
Atr8     2.809524
Atr9     2.892857
Atr10    2.773810
Atr11    3.214286
Atr12    2.940476
Atr13    3.095238
Atr14    2.880952
Atr15    2.940476
Atr16    2.821429
Atr17    3.166667
Atr18    2.976190
Atr19    3.178571
Atr20    2.880952
Atr21    2.654762
Atr22    2.452381
Atr23    2.773810
Atr24    2.785714
Atr25    2.952381
Atr26    2.809524
Atr27    2.678571
Atr28    2.559524
Atr29    2.928571
Atr30    2.821429
Atr31    3.440476
Atr32    3.416667
Atr33    3.357143
Atr34    3.273810
Atr35    3.273810
Atr36    3.214286
Atr37    3.583333
Atr38    3.404762
Atr39    3.642857
Atr40    3.571429
Atr41    3.547619
Atr42    3.333333
Atr43    3.476190
Atr44    3.380952
Atr45    3.285714
Atr46    3.166667
Atr47    3.321429
Atr48    3.452381
Atr49    3.511905
Atr50    3.500000
Atr51    3.357143
Atr52    3.488095
Atr53    3.321429
Atr54    3.369048
Class    1.000000
dtype: flo

In [12]:
df_mean = df.mean()
df_mean

Atr1     22.800466
Atr2     22.765199
Atr3     22.772519
Atr4     22.623335
Atr5     22.646668
Atr6     22.081388
Atr7     22.049102
Atr8     22.624914
Atr9     22.627100
Atr10    22.749750
Atr11    22.666915
Atr12    22.702699
Atr13    22.789214
Atr14    22.634169
Atr15    22.634661
Atr16    22.622590
Atr17    22.658432
Atr18    22.635456
Atr19    22.660275
Atr20    22.626625
Atr21    22.605048
Atr22    22.461699
Atr23    22.502976
Atr24    22.627019
Atr25    22.644936
Atr26    22.623585
Atr27    22.607135
Atr28    22.534209
Atr29    22.635804
Atr30    22.624817
Atr31    22.971311
Atr32    22.960284
Atr33    22.823886
Atr34    22.816314
Atr35    22.751602
Atr36    22.675537
Atr37    22.975536
Atr38    22.824203
Atr39    22.975966
Atr40    22.895828
Atr41    22.964485
Atr42    22.966607
Atr43    23.381791
Atr44    22.953180
Atr45    23.244844
Atr46    23.365591
Atr47    23.107179
Atr48    23.359816
Atr49    23.236742
Atr50    23.104313
Atr51    23.342089
Atr52    23.249273
Atr53    23.

In [13]:
amplitude = dfc1_mean - dfc0_mean
amplitude

Atr1     2.795127
Atr2     2.403931
Atr3     2.277132
Atr4     2.458749
Atr5     2.907254
Atr6     0.758859
Atr7     0.976467
Atr8     2.681617
Atr9     2.834718
Atr10    2.366833
Atr11    3.016611
Atr12    2.545127
Atr13    2.490587
Atr14    2.590255
Atr15    2.707918
Atr16    2.658638
Atr17    2.992248
Atr18    2.883167
Atr19    3.039037
Atr20    2.811185
Atr21    2.503599
Atr22    2.382614
Atr23    2.692414
Atr24    2.518272
Atr25    2.615172
Atr26    2.611849
Atr27    2.527409
Atr28    2.478128
Atr29    2.835548
Atr30    2.623754
Atr31    2.603267
Atr32    2.684109
Atr33    3.066445
Atr34    2.715670
Atr35    3.169158
Atr36    3.179402
Atr37    2.955426
Atr38    3.055925
Atr39    3.073090
Atr40    3.362126
Atr41    3.070875
Atr42    2.321705
Atr43    1.522702
Atr44    2.846069
Atr45    1.634551
Atr46    1.213178
Atr47    2.077243
Atr48    1.405869
Atr49    2.232835
Atr50    2.116279
Atr51    1.740864
Atr52    1.918328
Atr53    2.135382
Atr54    2.683001
Class    1.000000
dtype: flo

In [14]:
amplitude.sort_values(ascending=False)

Atr40    3.362126
Atr36    3.179402
Atr35    3.169158
Atr39    3.073090
Atr41    3.070875
Atr33    3.066445
Atr38    3.055925
Atr19    3.039037
Atr11    3.016611
Atr17    2.992248
Atr37    2.955426
Atr5     2.907254
Atr18    2.883167
Atr44    2.846069
Atr29    2.835548
Atr9     2.834718
Atr20    2.811185
Atr1     2.795127
Atr34    2.715670
Atr15    2.707918
Atr23    2.692414
Atr32    2.684109
Atr54    2.683001
Atr8     2.681617
Atr16    2.658638
Atr30    2.623754
Atr25    2.615172
Atr26    2.611849
Atr31    2.603267
Atr14    2.590255
Atr12    2.545127
Atr27    2.527409
Atr24    2.518272
Atr21    2.503599
Atr13    2.490587
Atr28    2.478128
Atr4     2.458749
Atr2     2.403931
Atr22    2.382614
Atr10    2.366833
Atr42    2.321705
Atr3     2.277132
Atr49    2.232835
Atr53    2.135382
Atr50    2.116279
Atr47    2.077243
Atr52    1.918328
Atr51    1.740864
Atr45    1.634551
Atr43    1.522702
Atr48    1.405869
Atr46    1.213178
Class    1.000000
Atr7     0.976467
Atr6     0.758859
dtype: flo

# **Conclusão**

Concluimos que as perguntas mais significativas são as que possuem maior amplitude (diferença entre a média das respostas por classe).

### **Para confirmar os resultados, poderíamos  repetir o experimento utilizando amostras**

Ao utilizar amostra com 1/5 da população, obtivemos um resultado diferente, porém com 4 questões em comum dentre as 6 com maior amplitude.

In [15]:
sample1 = dfc1.sample(34)
sample1.describe()

Atr1       Atr2       Atr3  ...      Atr53      Atr54  Class
count  34.000000  34.000000  34.000000  ...  34.000000  34.000000   34.0
mean    3.088235   2.764706   2.764706  ...   3.294118   3.323529    1.0
std     0.865768   0.923065   0.818677  ...   0.970143   1.272568    0.0
min     0.000000   0.000000   1.000000  ...   1.000000   0.000000    1.0
25%     3.000000   2.000000   2.000000  ...   3.000000   3.000000    1.0
50%     3.000000   3.000000   3.000000  ...   4.000000   4.000000    1.0
75%     4.000000   3.000000   3.000000  ...   4.000000   4.000000    1.0
max     4.000000   4.000000   4.000000  ...   4.000000   4.000000    1.0

[8 rows x 55 columns]

In [16]:
sample2 = dfc0.sample(34)
sample2.describe()

Atr1       Atr2       Atr3  ...      Atr53      Atr54  Class
count  34.000000  34.000000  34.000000  ...  34.000000  34.000000   34.0
mean    0.264706   0.676471   0.588235  ...   1.000000   0.588235    0.0
std     0.790428   0.911894   1.018545  ...   1.206045   0.820851    0.0
min     0.000000   0.000000   0.000000  ...   0.000000   0.000000    0.0
25%     0.000000   0.000000   0.000000  ...   0.000000   0.000000    0.0
50%     0.000000   0.000000   0.000000  ...   1.000000   0.000000    0.0
75%     0.000000   1.000000   1.000000  ...   2.000000   1.000000    0.0
max     4.000000   3.000000   4.000000  ...   4.000000   2.000000    0.0

[8 rows x 55 columns]

In [17]:
sample1_mean = sample1.mean()
sample2_mean = sample2.mean()
amplitude_sample = sample1_mean - sample2_mean
amplitude_sample.sort_values(ascending=False)

Atr40    3.411765
Atr35    3.294118
Atr36    3.205882
Atr38    3.205882
Atr41    3.176471
Atr33    3.117647
Atr19    3.117647
Atr39    3.088235
Atr17    3.029412
Atr37    3.000000
Atr5     2.911765
Atr29    2.911765
Atr18    2.882353
Atr11    2.882353
Atr1     2.823529
Atr20    2.794118
Atr31    2.794118
Atr34    2.764706
Atr15    2.735294
Atr9     2.735294
Atr54    2.735294
Atr16    2.705882
Atr26    2.705882
Atr23    2.705882
Atr42    2.647059
Atr44    2.647059
Atr30    2.647059
Atr25    2.647059
Atr14    2.617647
Atr28    2.617647
Atr32    2.617647
Atr12    2.588235
Atr27    2.529412
Atr21    2.529412
Atr24    2.500000
Atr8     2.500000
Atr22    2.411765
Atr47    2.411765
Atr10    2.382353
Atr13    2.382353
Atr4     2.352941
Atr53    2.294118
Atr49    2.205882
Atr3     2.176471
Atr52    2.147059
Atr2     2.088235
Atr50    1.970588
Atr43    1.764706
Atr45    1.617647
Atr51    1.588235
Atr46    1.382353
Atr48    1.294118
Class    1.000000
Atr6     0.882353
Atr7     0.794118
dtype: flo

In [18]:
dataset[(dataset['Class'] == 0)].describe()

Atr1       Atr2       Atr3  ...      Atr53      Atr54  Class
count  86.000000  86.000000  86.000000  ...  86.000000  86.000000   86.0
mean    0.395349   0.465116   0.639535  ...   1.186047   0.686047    0.0
std     0.898090   0.762313   0.919243  ...   1.183332   0.871262    0.0
min     0.000000   0.000000   0.000000  ...   0.000000   0.000000    0.0
25%     0.000000   0.000000   0.000000  ...   0.000000   0.000000    0.0
50%     0.000000   0.000000   0.000000  ...   1.000000   0.000000    0.0
75%     0.000000   1.000000   1.000000  ...   2.000000   1.000000    0.0
max     4.000000   3.000000   4.000000  ...   4.000000   4.000000    0.0

[8 rows x 55 columns]

# **Exercicio 2**

### Data: 25/Jan/2020

#### Criar um classificador utilizando o método Naive Bayes para identificar a probabilidade de um casal se divorciar baseado nas respostas das perguntas mais significativa, identificadas a partir do exercicio anterior levando em consideração a população total

## **Naive Bayes Classifier**


#### Utilizadando a biblioteca sklearn

### Divisão do Dataset contemplando as **seis perguntas** mais significativas 


In [19]:
df = dataset[['Atr33', 'Atr35', 'Atr36', 'Atr39', 'Atr40', 'Atr41', 'Class']]
df.head()

Atr33  Atr35  Atr36  Atr39  Atr40  Atr41  Class
0      1      0      1      3      3      2      1
1      2      0      2      2      4      2      1
2      1      1      1      3      3      3      1
3      2      1      1      4      4      2      1
4      1      0      0      2      1      0      1

**1o experimento:** utilizando uma divisão com 75% na base de treinamento e 25% na base de teste ou validação

In [0]:
X = df.drop('Class', axis = 1)
y = df['Class']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)

In [38]:
nbc = GaussianNB()
nbc.fit(X_train, y_train)
nbc_pred = nbc.predict(X_test)
score = nbc.score(X_train, y_train)
cr = classification_report(y_test, nbc_pred)
print('Accuracy Score: {0:.3f}'.format(score))
print ('Classification Report:\n %s' % cr)

Accuracy Score: 0.984
Classification Report:
               precision    recall  f1-score   support

           0       1.00      0.95      0.97        20
           1       0.96      1.00      0.98        23

    accuracy                           0.98        43
   macro avg       0.98      0.97      0.98        43
weighted avg       0.98      0.98      0.98        43



**2o experimento:** utilizando a média de 30 divisões com 75% na base de treinamento e 25% na base de teste ou validação

In [22]:
accuracy = []
media = 0
for i in range(30):
  X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)
  #random_state=i+1
  nbc = GaussianNB()
  nbc.fit(X_train, y_train)
  nbc_pred = nbc.predict(X_test)
  score = nbc.score(X_test, y_test)
  accuracy.append(score)
  media += score
print(media/30)
print(accuracy)

0.9635658914728681
[0.9767441860465116, 0.9534883720930233, 0.9534883720930233, 0.9767441860465116, 0.9534883720930233, 0.9534883720930233, 0.9767441860465116, 0.9534883720930233, 0.8837209302325582, 0.9767441860465116, 0.9767441860465116, 0.9534883720930233, 0.9534883720930233, 0.9534883720930233, 1.0, 0.9767441860465116, 0.9534883720930233, 0.9534883720930233, 0.9534883720930233, 0.9302325581395349, 0.9767441860465116, 0.9767441860465116, 0.9767441860465116, 0.9767441860465116, 1.0, 0.9767441860465116, 0.9534883720930233, 0.9767441860465116, 0.9767441860465116, 0.9534883720930233]


**3o experimento**: utilizando 5 divisões com rodízio manual, aplicando 80% na base de treinamento e 20% na base de teste ou validação

In [23]:
X_train = X[:136]
X_test = X[-34:]
y_train = y[:136]
y_test = y[-34:]
media_rodizio = 0
accuracy_rodizio = []

for i in range(5):
  nbc_rodizio = GaussianNB()
  nbc_rodizio.fit(X_train, y_train)
  nbc_rodizio_pred = nbc_rodizio.predict(X_test)
  score = nbc_rodizio.score(X_test, y_test)
  accuracy_rodizio.append(score)
  media_rodizio += score

  X_test_new = X_train[:34]
  X_train = X_train.append(X_test)
  X_train = X_train[-136:]
  X_test = X_test_new

  y_test_new = y_train[:34]
  y_train = y_train.append(y_test)
  y_train = y_train[-136:]
  y_test = y_test_new

print(media_rodizio/5)
print(accuracy_rodizio)

0.9588235294117646
[0.9117647058823529, 0.9117647058823529, 1.0, 0.9705882352941176, 1.0]


**4o experimento**: utilizando 5 divisões quebradas pelo método kFold, aplicando 80% na base de treinamento e 20% na base de teste ou validação

In [24]:
kf = KFold(n_splits=5)
i = 0
media = 0
accuracy = []

for train_index, test_index in kf.split(X):
  #print("TRAIN:", train_index, "TEST:", test_index)
  X_train, X_test = X[train_index[0]:train_index[-1]], X[test_index[0]:test_index[-1]]
  y_train, y_test = y[train_index[0]:train_index[-1]], y[test_index[0]:test_index[-1]]
  nbc = GaussianNB()
  nbc.fit(X_train, y_train)
  nbc_pred = nbc.predict(X_test)
  score = nbc.score(X_test, y_test)
  accuracy.append(score)
  media += score
  i += 1
print(media/i)
print(accuracy)

0.9636363636363636
[0.9090909090909091, 1.0, 1.0, 1.0, 0.9090909090909091]


Divisão do Dataset contemplando **as quatro perguntas mais significativas** que possuem intersecção com o resultado do experimento de **amostragem**

In [0]:
df2 = dataset[['Atr35', 'Atr36', 'Atr40', 'Atr41', 'Class']]
X2 = df2.drop('Class', axis = 1)
y2 = df2['Class']

In [26]:
accuracy = []
media = 0
for i in range(30):
  X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2, test_size=0.25)
  #random_state=i+1
  nbc2 = GaussianNB()
  nbc2.fit(X2_train, y2_train)
  nbc2_pred = nbc2.predict(X2_test)
  score2 = nbc2.score(X2_test, y2_test)
  accuracy.append(score2)
  media += score2
print(media/30)
print(accuracy)

0.9674418604651162
[0.9534883720930233, 0.9302325581395349, 0.9534883720930233, 1.0, 0.9534883720930233, 0.9302325581395349, 0.9534883720930233, 0.9534883720930233, 0.9767441860465116, 1.0, 0.9767441860465116, 0.9534883720930233, 0.9534883720930233, 1.0, 1.0, 0.9767441860465116, 0.9534883720930233, 0.9534883720930233, 0.9767441860465116, 0.9767441860465116, 0.9767441860465116, 0.9534883720930233, 0.9069767441860465, 0.9767441860465116, 1.0, 0.9534883720930233, 1.0, 0.9767441860465116, 0.9767441860465116, 0.9767441860465116]


In [27]:
score2 = nbc2.score(X2_train, y2_train)
cr2 = classification_report(y2_test, nbc2_pred)
print('Accuracy Score: {0:.3f}'.format(score2))
print ('Classification Report:\n %s' % cr2)

Accuracy Score: 0.984
Classification Report:
               precision    recall  f1-score   support

           0       0.95      1.00      0.97        19
           1       1.00      0.96      0.98        24

    accuracy                           0.98        43
   macro avg       0.97      0.98      0.98        43
weighted avg       0.98      0.98      0.98        43



In [0]:
df3 = dataset[['Atr11', 'Atr19', 'Atr33', 'Atr35', 'Atr36', 'Atr38', 'Atr39', 'Atr40', 'Atr41', 'Class']]
X3 = df3.drop('Class', axis = 1)
y3 = df3['Class']

In [29]:
accuracy = []
media = 0
for i in range(30):
  X3_train, X3_test, y3_train, y3_test = train_test_split(X3, y3, test_size=0.25)
  #random_state=i+1
  nbc3 = GaussianNB()
  nbc3.fit(X3_train, y3_train)
  nbc3_pred = nbc3.predict(X3_test)
  score3 = nbc3.score(X3_test, y3_test)
  accuracy.append(score3)
  media += score3
print(media/30)
print(accuracy)

0.9844961240310077
[0.9767441860465116, 1.0, 1.0, 1.0, 0.9767441860465116, 0.9767441860465116, 1.0, 1.0, 0.9767441860465116, 1.0, 0.9302325581395349, 1.0, 1.0, 1.0, 1.0, 0.9534883720930233, 0.9767441860465116, 0.9302325581395349, 0.9767441860465116, 1.0, 0.9767441860465116, 0.9767441860465116, 1.0, 0.9534883720930233, 0.9767441860465116, 1.0, 0.9767441860465116, 1.0, 1.0, 1.0]


In [30]:
score3 = nbc3.score(X3_train, y3_train)
cr3 = classification_report(y3_test, nbc3_pred)
print('Accuracy Score: {0:.3f}'.format(score3))
print ('Classification Report:\n %s' % cr3)

Accuracy Score: 0.984
Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        22
           1       1.00      1.00      1.00        21

    accuracy                           1.00        43
   macro avg       1.00      1.00      1.00        43
weighted avg       1.00      1.00      1.00        43



#**Conclusão** 

Definimos um primeiro modelo de classificação com acurácia de ~98% utilizando um subconjunto com apenas 6 perguntas

Definimos também um segundo modelo utilizando um subconjunto ainda menor com apenas 4 perguntas, com pouca perda de acurácia (~2%) em relação ao modelo anterior.

Identificamos que os modelos de classificação intermediários que utilizaram rodízio e a média da acurácia, obtiveram um resultado inferior devido a grande variação na acurácia para diferentes divisões das bases de treinamento e teste. PAra uma análise mais profunda desses experimentos intermediários, seria necessário calcular o desvio padrão.